### ISSUE TITLE
modify for failsafe? 

### ISSUE THREAD

**Issue by dantiel** [Nikkilae/PPM-reader#5]
Hi folks,
is there a way to add failsafe mechanism for this, when there are no new signal from receiver? i think this is a crucial feature.
best regards, d.

**Comment by supercamel** [comment]
I agree this is a crucial feature. Please add it, or at least a function to check when the last valid signal was received.

**Comment by dantiel** [comment]
actually its a one liner. this could also be just a usage example either incorporate default values in this lib:
```
if (micros() - ppm.microsAtLastPulse) > 1000) {
   rc_elevator = PPM1_DEFAULT_VALUE;
}
```

**Comment by Nikkilae** [comment]
Hi,

I agree that this would be a good feature to be added to the library. I will not push any changes to this repository until I can test my changes with actual hardware and I'm not sure when I get the chance to do that. @dantiel 's idea for implementing it seems reasonable to me. If you're in a hurry, you could write such a function to your local copy of the library until I push my solution.
```
// Public function for PPMReader:
unsigned long PPMReader::timeSinceLastPulse()
{
return micros() - microsAtLastPulse;
}
```
```
// In your code that reads the input:
if (ppmReader.timeSinceLastPulse() > tooMuchTime) {
    // Uh oh!
}
```
I'm hesitant to define a length of time into the library after which a failsafe procedure would be needed. As I've understood it, different hardware will have varying specifications as to what that number would be.

**Comment by supercamel** [comment]
Perhaps you could do a function called 'setFailsafe' or something, which accepts a timeout parameter and an array of fail safe positions for each channel. That would be convenient, I think.

**Comment by dantiel** [comment]
keep it  slim,  because program space  on  arduino might become  small very fast using many functions. a good way would be rewrite  the library  so  it uses only `#define` instead of other variables, then those can be defined in our main sheet, and use  `#ifdef`. possible candidates for those  constants may be `FAILSAFE_THRESHOLD`, `PPM_CHANNEL_COUNT`, `PPM1_FAILSAFE`,  `PPM2_FAILSAFE` ...

**Comment by dantiel** [comment]
an array would also be clean. but to saving space `define` works definitely well

**Comment by supercamel** [comment]
You could achieve the same effect using templates. 


```
template <const int* FAILSAFES, const int NUM_CHANNELS, const int TIMEOUT> class ppm {
  public:
    void printFailsafes() {
      for(int i = 0; i < NUM_CHANNELS; i++) {
        Serial.print(FAILSAFES[i]);
        Serial.print(" ");
      }
      Serial.println();
    }
};

const int failsafes[] = {1500, 1500, 1000, 1500};
const int numChannels = 4;
const int timeout = 100;

ppm<failsafes, numChannels, timeout> p;
```


This just as lean as using #define but has a more C++ aesthetic, I suppose.
